numpy/core/src/multiarray/scalartypes.c.src

```diff
@@ -373,31 +374,62 @@ gentype_nonzero_number(PyObject *m1)
 static PyObject *
 gentype_str(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
-
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    static PyObject *fallback = NULL;
+    npy_cache_import("numpy.core.arrayprint", "_scalar_fallback", &fallback);
+    if (fallback == NULL) {
+        return NULL;
     }
-    return ret;
+    return PyObject_CallFunction(fallback, "O", self);
 }
 
 
 static PyObject *
 gentype_repr(PyObject *self)
 {
-    PyObject *arr, *ret = NULL;
+    PyObject  *dtype_type, *dtype_name, *item, *item_repr;
 
-    arr = PyArray_FromScalar(self, NULL);
-    if (arr != NULL) {
-        /* XXX: Why are we using str here? */
-        ret = PyObject_Str((PyObject *)arr);
-        Py_DECREF(arr);
+    dtype_type = (PyObject*)PyArray_DescrFromScalar(self)->typeobj;
+    dtype_name = PyObject_GetAttrString(dtype_type, "__name__");
+    if (dtype_name == NULL) {
+        /* this should never happen since types always have __name__ */
+        PyErr_Format(PyExc_RuntimeError, "missing __name__ attribute");
+        return NULL;
```

Comment:
CPython just leaves the error as is internally: https://github.com/python/cpython/blob/ba85d69a3e3610bdd05f0dd372cf4ebca178c7fb/Python/_warnings.c#L315